source/winAPI/sessionTracking.py

```diff
@@ -106,110 +90,105 @@ class WindowsTrackedSession(enum.IntEnum):
 
 
-_toggleWindowsSessionStatePair: Dict[WindowsTrackedSession, WindowsTrackedSession] = {
-	WindowsTrackedSession.CONSOLE_CONNECT: WindowsTrackedSession.CONSOLE_DISCONNECT,
-	WindowsTrackedSession.CONSOLE_DISCONNECT: WindowsTrackedSession.CONSOLE_CONNECT,
-	WindowsTrackedSession.REMOTE_CONNECT: WindowsTrackedSession.REMOTE_DISCONNECT,
-	WindowsTrackedSession.REMOTE_DISCONNECT: WindowsTrackedSession.REMOTE_CONNECT,
-	WindowsTrackedSession.SESSION_LOGON: WindowsTrackedSession.SESSION_LOGOFF,
-	WindowsTrackedSession.SESSION_LOGOFF: WindowsTrackedSession.SESSION_LOGON,
-	WindowsTrackedSession.SESSION_LOCK: WindowsTrackedSession.SESSION_UNLOCK,
-	WindowsTrackedSession.SESSION_UNLOCK: WindowsTrackedSession.SESSION_LOCK,
-	WindowsTrackedSession.SESSION_CREATE: WindowsTrackedSession.SESSION_TERMINATE,
-	WindowsTrackedSession.SESSION_TERMINATE: WindowsTrackedSession.SESSION_CREATE,
-}
-"""
-Pair of WindowsTrackedSession, where each key has a value of the opposite state.
-e.g. SESSION_LOCK/SESSION_UNLOCK.
-"""
+class _CurrentSessionState(AutoPropertyObject):
+
+	def __init__(self) -> None:
+		self._windowsLockStateIsUnknown = False
+		"""
+		Track if any 'Unknown' Value when querying the Session Lock status has been encountered.
+		Resets when a successful Windows lock state query completes.
+		"""
+		self._prev_isWindowsLockedValue = False
+		"""Before NVDA initialization, assume Windows is unlocked.
+		This means if Windows is locked, NVDA processes the state change correctly.
+		"""
+		super().__init__()
+
+	_cache_isWindowsLocked = True
+
+	# typing information for auto property _get_isWindowsLocked
+	isWindowsLocked: bool
+
+	def _get_isWindowsLocked(self) -> bool:
+		# Reset this at the start of every cache cycle.
+		self._windowsLockStateIsUnknown = False
+		if _isSecureDesktop():
+			# If this is the Secure Desktop,
+			# we are in secure mode and on a secure screen,
+			# e.g. on the sign-in screen.
+			# _isSecureDesktop may also return True on the lock screen before a user has signed in.
+			# For more information, refer to devDocs/technicalDesignOverview.md 'Logging in secure mode'
+			# and the following userGuide sections:
+			# - SystemWideParameters (information on the serviceDebug parameter)
+			# - SecureMode and SecureScreens
+			isWindowsLocked = False
+		else:
+			try:
+				isWindowsLocked = _isWindowsLocked_checkViaSessionQuery()
+			except Exception as e:
+				self._recordLockStateTrackingFailure(e)  # Report error repeatedly, attention is required.
+				##
+				# For security it would be best to treat unknown as locked.
+				# However, this would make NVDA unusable.
+				# Instead, the user should be warned via UI, and allowed to mitigate the impact themselves.
+				# See usage of L{warnSessionLockStateUnknown}.
+				isWindowsLocked = False
+
+		if self._prev_isWindowsLockedValue != isWindowsLocked:
+			from utils.security import postSessionLockStateChanged
+			postSessionLockStateChanged.notify(isNowLocked=isWindowsLocked)

```

Comment:
This is concerning; we can't guarantee that the lock state changed notification will happen, or the timing. It relies on other code checking for lock state. It's an indirect mechanism. An alternative would be to add another class, a `lockStateChangedObserver` that gets updated once per core cycle (like pumpAll for other modules) and handles the comparison of `_prev_isWindowsLockedValue` with the latest value.